From 17d2781b431239922798bace0cf47e85de219dce Mon Sep 17 00:00:00 2001
From: Nizan Zorea <nzorea@marvell.com>
Date: Mon, 21 Mar 2016 15:06:16 +0200
Subject: [PATCH 0969/1240] rfu: add rfu-dump support for apn-a0

update the rfu prints to match ATF driver

Change-Id: I9e936b73bf600d621a6ac84e3bc220a4a970aa41
Signed-off-by: Nizan Zorea <nzorea@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/28409
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Haim Boot <hayim@marvell.com>
---
 arch/arm/cpu/armv8/armada8k/mvebu_rfu.c | 81 ++++++++++++++++++++++++++++++---
 arch/arm/dts/apn-806-a0.dtsi            |  6 +++
 2 files changed, 81 insertions(+), 6 deletions(-)

diff --git a/arch/arm/cpu/armv8/armada8k/mvebu_rfu.c b/arch/arm/cpu/armv8/armada8k/mvebu_rfu.c
index 5a81e91..3a8b76a 100644
--- a/arch/arm/cpu/armv8/armada8k/mvebu_rfu.c
+++ b/arch/arm/cpu/armv8/armada8k/mvebu_rfu.c
@@ -31,13 +31,14 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#ifdef CONFIG_AP806_Z_SUPPORT
 enum rfu_target_ids {
-	BOOTROM_TID     = 0x0,
-	STM_TID         = 0x1,
-	SPI_TID         = 0x2,
-	PCIE_REGS_TID   = 0x3,
-	IHBPHY_TID      = 0x4,
-	PCIE_PORT_TID   = 0x5,
+	BOOTROM_TID      = 0x0,
+	STM_TID          = 0x1,
+	SPI_TID          = 0x2,
+	PCIE_REGS_TID    = 0x3,
+	IHBPHY_TID       = 0x4,
+	PCIE_PORT_TID    = 0x5,
 	RFU_MAX_TID
 };
 
@@ -104,3 +105,71 @@ void dump_rfu(void)
 
 	return;
 }
+
+#else
+
+/* AP registers */
+#define RFU_WIN_ALR_OFFSET(win)         (0x0 + (0x10 * win))
+#define RFU_WIN_AHR_OFFSET(win)         (0x8 + (0x10 * win))
+#define RFU_WIN_CR_OFFSET(win)          (0xC + (0x10 * win))
+
+enum rfu_target_ids {
+	IHB_0_TID        = 0x0,
+	IHB_1_TID        = 0x1,
+	IHB_2_TID        = 0x2,
+	PIDI_TID         = 0x3,
+	SPI_TID          = 0x4,
+	STM_TID          = 0x5,
+	BOOTROM_TID      = 0x6,
+	RFU_MAX_TID
+};
+
+void dump_rfu(void)
+{
+	u32 trgt_id, win_id;
+	u32 alr, ahr, rfu_max_win;
+	u64 start, end;
+	const void *blob = gd->fdt_blob;
+	void __iomem *rfu_base;
+	char *rfu_target_name[RFU_MAX_TID] = {"IHB-0    ", "IHB-1    ", "IHB-2    ", "PIDI     ",
+						"SPI      ", "STM      ", "BootRoom "};
+	int node;
+
+	/* Get address decoding node from the FDT blob */
+	node = fdt_node_offset_by_compatible(blob, -1, fdtdec_get_compatible(COMPAT_MVEBU_RFU));
+	if (node < 0) {
+		error("No RFU address decoding node found in FDT blob\n");
+		return;
+	}
+
+	/* Get the base address of the address decoding MBUS */
+	rfu_base = (void *)fdt_get_regs_offs(blob, node, "reg");
+
+	/* Get the  max windows fot rfu */
+	rfu_max_win = fdtdec_get_int(blob, node, "max-win", -1);
+
+	/* Dump all RFU windows */
+	printf("bank  target     start              end\n");
+	printf("----------------------------------------------------\n");
+	for (win_id = 0; win_id < rfu_max_win; win_id++) {
+		alr = readl(rfu_base + RFU_WIN_ALR_OFFSET(win_id));
+		if (alr & WIN_ENABLE_BIT) {
+			alr &= ~WIN_ENABLE_BIT;
+			/* in case this is BOOTROM window */
+			if (win_id == 0) {
+				ahr = alr;
+				trgt_id = BOOTROM_TID;
+			} else {
+				ahr = readl(rfu_base + RFU_WIN_AHR_OFFSET(win_id));
+				trgt_id = readl(rfu_base + RFU_WIN_CR_OFFSET(win_id));
+			}
+			start = ((u64)alr << ADDRESS_SHIFT);
+			end = (((u64)ahr + 0x10) << ADDRESS_SHIFT);
+			printf("rfu   %s  0x%016llx 0x%016llx\n", rfu_target_name[trgt_id], start, end);
+		}
+	}
+	printf("rfu   PIDI-port  - all other IO transactions\n");
+
+	return;
+}
+#endif  /* CONFIG_AP806_Z_SUPPORT */
diff --git a/arch/arm/dts/apn-806-a0.dtsi b/arch/arm/dts/apn-806-a0.dtsi
index efd0770..e298614 100644
--- a/arch/arm/dts/apn-806-a0.dtsi
+++ b/arch/arm/dts/apn-806-a0.dtsi
@@ -79,6 +79,12 @@
 				reg = <0x4000 0x50>;
 				max-win = <8>;
 			};
+			rfu {
+
+				compatible = "marvell,mvebu-rfu";
+				reg = <0x6f0000 0x28>;
+				max-win = <7>;
+			};
 
 			thermal: thermal@6f8084 {
 				compatible = "marvell,thermal-sensor-v2";
-- 
1.9.1

